File: Android-My-Nanny-old-version/app/src/main/java/com/vald3nir/mynanny/views/register/RegisterActivity.java
package com.vald3nir.mynanny.views.register;

import android.os.Bundle;
import android.widget.EditText;

import com.vald3nir.mynanny.R;
import com.vald3nir.mynanny.views.custom.CustomActivity;

import static com.vald3nir.mynanny.utils.AppUtils.getValue;

public class RegisterActivity extends CustomActivity {

    private EditText edtEmail, edtPassword, edtConfirmPassword;
    private RegisterDelegate delegate;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);
        delegate = new RegisterDelegate(this);
        initViews();
    }

    private void initViews() {
        edtEmail = findViewById(R.id.edt_email);
        edtPassword = findViewById(R.id.edt_password);
        edtConfirmPassword = findViewById(R.id.edt_confirm_password);
        findViewById(R.id.btn_register).setOnClickListener(v -> register());
    }

    private void register() {
        delegate.register(this, getValue(edtEmail), getValue(edtPassword), getValue(edtConfirmPassword));
    }
}

File: app/src/main/java/com/vald3nir/my_nanny/presentation/register/RegisterActivity.kt
package com.vald3nir.my_nanny.presentation.register

import android.os.Bundle
import android.view.inputmethod.EditorInfo
import androidx.lifecycle.Observer
import com.vald3nir.my_nanny.common.core.BaseActivity
import com.vald3nir.my_nanny.common.extensions.afterTextChanged
import com.vald3nir.my_nanny.databinding.ActivityRegisterBinding
import org.koin.androidx.viewmodel.ext.android.viewModel

class RegisterActivity : BaseActivity() {

    private val viewModel: RegisterViewModel by viewModel()
    private lateinit var binding: ActivityRegisterBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityRegisterBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel.view = this
        setupObservers()
    }

    private fun setupObservers() {

        binding.apply {

            btnBack.setOnClickListener { onBackPressed() }
            btnRegister.setOnClickListener { register() }

            edtEmail.afterTextChanged { registerDataChanged() }

            edtPassword.apply { afterTextChanged { registerDataChanged() } }

            edtConfirmPassword.apply {
                afterTextChanged { registerDataChanged() }
                setOnEditorActionListener { _, actionId, _ ->
                    when (actionId) {
                        EditorInfo.IME_ACTION_DONE ->
                            register()
                    }
                    false
                }
            }
        }

        viewModel.registerFormState.observe(this@RegisterActivity, Observer {
            val loginState = it ?: return@Observer
            binding.btnRegister.isEnabled = loginState.isDataValid

            if (loginState.usernameError != null) {
                binding.edtEmail.error = getString(loginState.usernameError)
            }
            if (loginState.passwordError != null) {
                binding.edtPassword.error = getString(loginState.passwordError)
            }
            if (loginState.passwordNotEquals != null) {
                binding.edtConfirmPassword.error = getString(loginState.passwordNotEquals)
            }
        })
    }

    private fun ActivityRegisterBinding.registerDataChanged() {
        viewModel.registerDataChanged(
            edtEmail.text.toString(),
            edtPassword.text.toString(),
            edtConfirmPassword.text.toString()
        )
    }

    private fun ActivityRegisterBinding.register() {
        viewModel.register(
            email = edtEmail.text.toString(),
            password = edtPassword.text.toString()
        )
    }
}
